pkg/allocateip/allocateip.go

```diff
@@ -176,23 +181,25 @@ func ensureHostTunnelAddress(ctx context.Context, c client.Interface, nodename s
 		} else if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
 			// The tunnel address is not assigned, reassign it.
 			logCtx.WithField("currentAddr", addr).Info("Current address is not assigned, assign a new one")
+
+			// Defensively release any IP addresses with this handle. This covers a theoretical case
+			// where the node object has lost its reference to its correct IP, but the allocation still exists
+			// in IPAM. For example, if the node object was manually edited.
+			release = true
 		} else {
 			// Failed to get assignment attributes, datastore connection issues possible, panic
 			logCtx.WithError(err).Panicf("Failed to get assignment attributes for CIDR '%s'", addr)
 		}
 	}
 
 	if release {
-		logCtx.WithField("IP", addr).Info("Release old tunnel address")
-		ipAddr := net.ParseIP(addr)
-		if ipAddr == nil {
-			logCtx.WithError(err).Fatalf("Failed to parse the CIDR '%s'", addr)
-		}
-
-		ipsToRelease := []net.IP{*ipAddr}
-		_, err := c.IPAM().ReleaseIPs(ctx, ipsToRelease)
-		if err != nil {
-			logCtx.WithField("IP", ipAddr.String()).WithError(err).Fatal("Error releasing address")
+		logCtx.WithField("IP", addr).Info("Release any old tunnel addresses")
+		handle, _ := generateHandleAndAttributes(nodename, vxlan)
+		if err := c.IPAM().ReleaseByHandle(ctx, handle); err != nil {
+			if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+				logCtx.WithError(err).Fatal("Failed to release old addresses")
+			}
+			// No existing allocations for this node.
```

Comment:
Should we not keep the old code as well?  I.e., both release by IP (`addr`) and release by handle?

Reply:
(If so, should we also have a UT for the old release by IP case?)

Reply:
I guess I was assuming that by the time we got here, either:
- The address was originally assigned with a handle, and so we're OK
- The address was originally assigned without a handle, but was corrected to have one so we're OK
- The address was originally assigned without a handle, but no longer was on the node object so we don't have an IP to use anyway.
- The address was originally assigned without a handle, but the node object was modified by the user to something wrong. I don't think its safe to just unassign whatever address was in use, since it might have been modified to match a pod IP, for example, and we wouldn't want to blindly release that address.

Does that make sense? Or am I missing a case?

Reply:
Just one remaining concern:

> The address was originally assigned without a handle, but was corrected to have one so we're OK

Will the correction code (in the previous PR) have run before the code here?

Reply:
Actually, OK, I can see that we're good for all of the cases in the code above that set `release = true`.